drivers/acpi/acpi.c

```diff
@@ -0,0 +1,554 @@
+
+	return acpi_get_irq_table(bus, bus_name, rt_table, rt_size);
+}
+
+ACPI_RESOURCE *acpi_resource_parse(const struct device *dev, ACPI_RESOURCE *res, int res_type)
+{
+	do {
+		if (!res->Length) {
+			LOG_DBG("Error: zero length found!\n");
+			break;
+		} else if (res->Type == res_type) {
+			break;
+		}
+		res = ACPI_NEXT_RESOURCE(res);
+	} while (res->Type != ACPI_RESOURCE_TYPE_END_TAG);
+
+	if (res->Type == ACPI_RESOURCE_TYPE_END_TAG) {
+		return NULL;
+	}
+
+	return res;
+}
+
+static int acpi_res_type(ACPI_RESOURCE *res)
+{
+	int type;
+
+	switch (res->Type) {
+	case ACPI_RESOURCE_TYPE_IO:
+		type = ACPI_RESOURCE_TYPE_IO;
+		break;
+	case ACPI_RESOURCE_TYPE_FIXED_IO:
+		type = ACPI_RESOURCE_TYPE_FIXED_IO;
+		break;
+	case ACPI_RESOURCE_TYPE_MEMORY24:
+		type = ACPI_RESOURCE_TYPE_MEMORY24;
+		break;
+	case ACPI_RESOURCE_TYPE_MEMORY32:
+		type = ACPI_RESOURCE_TYPE_MEMORY32;
+		break;
+	case ACPI_RESOURCE_TYPE_FIXED_MEMORY32:
+		type = ACPI_RESOURCE_TYPE_FIXED_MEMORY32;
+		break;
+	case ACPI_RESOURCE_TYPE_ADDRESS16:
+		type = ACPI_RESOURCE_TYPE_ADDRESS16;
+		break;
+	case ACPI_RESOURCE_TYPE_ADDRESS32:
+		type = ACPI_RESOURCE_TYPE_ADDRESS32;
+		break;
+	case ACPI_RESOURCE_TYPE_ADDRESS64:
+		type = ACPI_RESOURCE_TYPE_ADDRESS64;
+		break;
+	case ACPI_RESOURCE_TYPE_EXTENDED_ADDRESS64:
+		type = ACPI_RESOURCE_TYPE_EXTENDED_ADDRESS64;
+		break;
+	default:
+		type = ACPI_RESOURCE_TYPE_MAX;
+	}
+
+	return type;
+}
+
+int acpi_device_type_get(const struct device *dev, ACPI_RESOURCE *res)
+{
+	int type = ACPI_RESOURCE_TYPE_MAX;
+
+	do {
+		if (!res->Length) {
+			LOG_ERR("Error: zero length found!\n");
+			break;
+		}
+		type = acpi_res_type(res);
+		if (type != ACPI_RESOURCE_TYPE_MAX) {
+			break;
+		}
+		res = ACPI_NEXT_RESOURCE(res);
+	} while (res->Type != ACPI_RESOURCE_TYPE_END_TAG);
+
+	return type;
+}
+
+struct acpi_dev *acpi_device_get(const struct device *dev, char *hid, int inst)
+{
+	struct acpi_dev *child_dev;
+	int i = 0, inst_id;
+	struct acpi *bus = dev->data;
+
+	do {
+		child_dev = &bus->child_dev[i];
+		if (!child_dev->path) {
+			LOG_DBG("NULL device path found\n");
+			continue;
+		}
+
+		if (!child_dev->res_lst || !child_dev->dev_info ||
+		    !child_dev->dev_info->HardwareId.Length) {
+			continue;
+		}
+
+		if (!strcmp(hid, child_dev->dev_info->HardwareId.String)) {
+			if (child_dev->dev_info->UniqueId.Length) {
+				inst_id = atoi(child_dev->dev_info->UniqueId.String);
+				if (inst_id == inst) {
+					return child_dev;
+				}
+			} else {
+				return child_dev;
+			}
+		}
+	} while (i++ < bus->num_dev);
+
+	return NULL;
+}
+
+struct acpi_dev *acpi_device_by_index_get(const struct device *dev, int index)
+{
+	struct acpi *bus = dev->data;
+
+	return index < bus->num_dev ? &bus->child_dev[index] : NULL;
+}
+
+static struct acpi bus_ctx = {
+	.status = AE_NOT_CONFIGURED,
+};
+
+/* TODO: need to enable initialization levels as PRE_KERNEL_1. */
+DEVICE_DEFINE(acpi, "acpi", acpi_init, NULL, &bus_ctx, NULL, POST_KERNEL,
```

Comment:
Actually, acpi is not a device per se. 

So, either you call SYS_INIT() or, better because this would not end up dependent on a level/prio, make acpi_init() public and let the arch to call it at the right place (currently, only x86 would). (perhaps that would mean also acpi should not be in drivers/ but in lib/ or something like that)

Reply:
SYS_INIT is also have level/prio and order of invoked based on it.

Reply:
updated with SYS_INIT

Reply:
> SYS_INIT is also have level/prio and order of invoked based on it.

Thus the better proposal, on the same comment: making acpi_init() public and calling it directly in arch/x86/ relevant place.

Reply:
> Thus the better proposal, on the same comment: making acpi_init() public and calling it directly in arch/x86/ relevant place.

This sounds like a good approach for me as well, i.e. do the init from the arch code. Btw, is `drivers/acpi/` the right location? Or would `lib/acpi/` or `subsys/acpi/` make more sense?

Reply:
ACPI is mostly deal with platform HW and devices. So my view is that it should be part of drivers since it mostly use by drivers. For other OS like Linux and Windows is exposed as drivers.

Reply:
> > SYS_INIT is also have level/prio and order of invoked based on it.
> 
> Thus the better proposal, on the same comment: making acpi_init() public and calling it directly in arch/x86/ relevant place.

For CPU configurations etc. acpica provided initialization sequence for early ACPI table access etc. before doing full acpica initialization. This can be done as part of x86 arch porting update for using acpica APIs and currently it is enabled only for device drivers.

Reply:
> ACPI is mostly deal with platform HW and devices. So my view is that it should be part of drivers since it mostly use by drivers. For other OS like Linux and Windows is exposed as drivers.

It is _used_ by drivers but acpi itself is not a device and thus the acpica is more like a lib or a subsys.

Reply:
moved acpi under lib/acpi.